kale/embed/_mpca_torch.py

```diff
@@ -0,0 +1,230 @@
+                self.factors[i - 1] = (eig_vectors[:, idx_sorted][:, : self.shape_out[i - 1]]).T
+
+        tensor_pc_vec = self.forward(x, return_vector=True, sort_by_var=False)
+        # return_vector the tensor principal components
+        # diagonal of the covariance of vectorised features
+        tensor_pc_cov_diag = torch.diag(torch.matmul(tensor_pc_vec.T, tensor_pc_vec))
+        idx_order = (-1 * tensor_pc_cov_diag).argsort()
+
+        self.idx_order = idx_order
+
+        return self
+
+    def training_step(self, train_batch):
+        """Update the model parameters for a training batch"""
+        x, y = train_batch
+        if len(self.factors) == 0:
+            self._init_model(x)
+        self._update_model(x)
+        return self
+
+    def forward(self, x, return_vector=False, sort_by_var=True, n_components=None):
+        """Perform dimension reduction on x
+
+        Args:
+            x (tensor):
+            # reduce_mean (bool): Whether reduce the mean estimated from training data. Defaults to True.
+            return_vector (bool): Whether vectorize the tensor principal components. Defaults to False.
+            sort_by_var (bool): Whether sort features by variance in descending order. Defaults to True,
+                applies only when return_vector=True.
+            n_components (int): Number of components to keep. Applies only when return_vector=True.
+                Defaults to None.
+
+        Returns:
+            tensor
+                tensor principal components, shape (n_samples, P_1, P_2, ..., P_N) if return_vector==False.
+                If return_vector==True, features will be vectorized and sorted based on their explained
+                variance ratio, shape (n_samples, P_1 * P_2 * ... * P_N) if self.n_components is None,
+                and shape (n_samples, n_components) if self.n_component is not None.
+        """
+        _check_dim_shape(x, self.n_dim, self.shape_in)
+
+        # if reduce_mean:
+        #     x -= self.mean_
+
+        # tensor principal components
+        tensor_pc = multi_mode_dot(x, self.factors, modes=[m for m in range(1, self.n_dim)])
+
+        if return_vector:
+            tensor_pc = tl.unfold(tensor_pc, mode=0)
+            if sort_by_var:
+                tensor_pc = tensor_pc[:, self.idx_order]
+            if isinstance(n_components, int):
+                all_comp = int(torch.prod(torch.Tensor(self.shape_out)))
+                if n_components > all_comp:
+                    warnings.warn(
+                        "n_components can not exceed the multiplication of output "
+                        "tensor shapes, the maximum number of components will be applied."
+                    )
+                    n_components = all_comp
+                if not sort_by_var:
+                    tensor_pc = tensor_pc[:, self.idx_order]
+                tensor_pc = tensor_pc[:, :n_components]
+
+        return tensor_pc
+
+    def restore(self, x):
```

Comment:
Did you see the word "restore" used in this context? To me "restore" (bring back) does not have the meaning of transformation.
To be more classical/traditional, "reconstruct" is the proper word. We minimize reconstruction errors.
To be more modern, you could consider "decode" because PCA is basically a linear autoencoder (PCA: encoder; inverse: decoder).

Reply:
Will change to reconstruct